File: contracts/deploy.py
"""
ConsentVault - Contract Compilation and Deployment
Compiles PyTeal to TEAL and deploys to Algorand TestNet
"""

import os
import base64
from algosdk import account, mnemonic
from algosdk.v2client import algod
from algosdk.transaction import ApplicationCreateTxn, StateSchema, OnComplete, wait_for_confirmation

from approval import approval_program
from clear import clear_state_program

try:
    from pyteal import compileTeal, Mode
except ImportError:
    print("PyTeal not installed. Run: pip install pyteal")
    exit(1)


def compile_contracts():
    """Compile PyTeal programs to TEAL"""
    print("Compiling PyTeal to TEAL...")
    
    approval_teal = compileTeal(approval_program(), mode=Mode.Application, version=8)
    clear_teal = compileTeal(clear_state_program(), mode=Mode.Application, version=8)
    
    # Save compiled TEAL
    with open("consent_approval.teal", "w") as f:
        f.write(approval_teal)
    print(f"  - consent_approval.teal ({len(approval_teal)} bytes)")
    
    with open("consent_clear.teal", "w") as f:
        f.write(clear_teal)
    print(f"  - consent_clear.teal ({len(clear_teal)} bytes)")
    
    # Also save combined consent.teal for reference
    combined = f"""// ConsentVault Smart Contract
// DPDP Act Compliant Consent Management
// Compiled from PyTeal - TEAL v8

// ============================================
// APPROVAL PROGRAM
// ============================================
{approval_teal}

// ============================================
// CLEAR STATE PROGRAM  
// ============================================
{clear_teal}
"""
    with open("consent.teal", "w") as f:
        f.write(combined)
    print(f"  - consent.teal (combined reference)")
    
    return approval_teal, clear_teal


def get_algod_client():
    """Initialize Algorand client"""
    algod_address = os.environ.get("ALGOD_SERVER", "https://testnet-api.algonode.cloud")
    algod_token = os.environ.get("ALGOD_TOKEN", "")
    
    return algod.AlgodClient(algod_token, algod_address)


def compile_program(client, source_code):
    """Compile TEAL source to bytecode"""
    compile_response = client.compile(source_code)
    return base64.b64decode(compile_response["result"])


def deploy_contract(approval_teal, clear_teal):
    """Deploy contract to Algorand TestNet"""
    print("\nDeploying to Algorand TestNet...")
    
    # Get deployer account
    deployer_mnemonic = os.environ.get("DEPLOYER_MNEMONIC")
    if not deployer_mnemonic:
        print("ERROR: DEPLOYER_MNEMONIC environment variable required")
        print("\nTo generate a test account:")
        print("  from algosdk import account, mnemonic")
        print("  sk, addr = account.generate_account()")
        print("  print(mnemonic.from_private_key(sk))")
        print("\nThen fund it at: https://bank.testnet.algorand.network/")
        return None
    
    deployer_sk = mnemonic.to_private_key(deployer_mnemonic)
    deployer_addr = account.address_from_private_key(deployer_sk)
    print(f"Deployer: {deployer_addr}")
    
    # Initialize client
    client = get_algod_client()
    
    # Check balance
    account_info = client.account_info(deployer_addr)
    balance = account_info.get("amount", 0) / 1_000_000
    print(f"Balance: {balance:.4f} ALGO")
    
    if balance < 0.5:
        print("ERROR: Insufficient balance. Need at least 0.5 ALGO")
        print("Fund at: https://bank.testnet.algorand.network/")
        return None
    
    # Compile programs
    approval_program = compile_program(client, approval_teal)
    clear_program = compile_program(client, clear_teal)
    
    print(f"Approval program: {len(approval_program)} bytes")
    print(f"Clear program: {len(clear_program)} bytes")
    
    # Define schema
    global_schema = StateSchema(num_uints=1, num_byte_slices=1)
    local_schema = StateSchema(num_uints=1, num_byte_slices=16)
    
    # Get suggested params
    params = client.suggested_params()
    
    # Create application transaction
    txn = ApplicationCreateTxn(
        sender=deployer_addr,
        sp=params,
        on_complete=OnComplete.NoOpOC,
        approval_program=approval_program,
        clear_program=clear_program,
        global_schema=global_schema,
        local_schema=local_schema,
    )
    
    # Sign and send
    signed_txn = txn.sign(deployer_sk)
    tx_id = client.send_transaction(signed_txn)
    print(f"\nTransaction ID: {tx_id}")
    print("Waiting for confirmation...")
    
    # Wait for confirmation
    confirmed = wait_for_confirmation(client, tx_id, 4)
    app_id = confirmed["application-index"]
    
    print(f"\n{'='*50}")
    print(f"SUCCESS! Contract deployed")
    print(f"{'='*50}")
    print(f"Application ID: {app_id}")
    print(f"Explorer: https://testnet.algoexplorer.io/application/{app_id}")
    print(f"\nAdd to .env.local:")
    print(f"NEXT_PUBLIC_APP_ID={app_id}")
    
    # Save deployment info
    import json
    deployment_info = {
        "app_id": app_id,
        "deployer": deployer_addr,
        "tx_id": tx_id,
        "network": "testnet",
    }
    
    with open("../deployment.json", "w") as f:
        json.dump(deployment_info, f, indent=2)
    print(f"\nDeployment info saved to deployment.json")
    
    return app_id


def main():
    print("=" * 50)
    print("ConsentVault Contract Deployment")
    print("=" * 50)
    print()
    
    # Compile
    approval_teal, clear_teal = compile_contracts()
    
    # Deploy if mnemonic provided
    if os.environ.get("DEPLOYER_MNEMONIC"):
        deploy_contract(approval_teal, clear_teal)
    else:
        print("\nCompilation complete!")
        print("To deploy, set DEPLOYER_MNEMONIC environment variable")


if __name__ == "__main__":
    main()

File: scripts/deploy.js
/**
 * ConsentVault - Deployment Script
 * Deploys the smart contract to Algorand TestNet
 * 
 * Usage: node scripts/deploy.js
 * 
 * Required env vars:
 * - ALGOD_TOKEN: Algorand node API token
 * - ALGOD_SERVER: Algorand node URL
 * - DEPLOYER_MNEMONIC: 25-word mnemonic for deployment account
 */

import algosdk from 'algosdk';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Configuration
const config = {
  algodToken: process.env.ALGOD_TOKEN || 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
  algodServer: process.env.ALGOD_SERVER || 'https://testnet-api.algonode.cloud',
  algodPort: process.env.ALGOD_PORT || '',
};

async function deployContract() {
  console.log('🚀 ConsentVault Deployment Script');
  console.log('==================================\n');

  // Initialize Algod client
  const algodClient = new algosdk.Algodv2(
    config.algodToken,
    config.algodServer,
    config.algodPort
  );

  // Get deployer account
  const mnemonic = process.env.DEPLOYER_MNEMONIC;
  if (!mnemonic) {
    console.error('❌ DEPLOYER_MNEMONIC environment variable is required');
    console.log('\nTo generate a new account for testing:');
    console.log('const account = algosdk.generateAccount();');
    console.log('console.log(algosdk.secretKeyToMnemonic(account.sk));');
    process.exit(1);
  }

  const deployer = algosdk.mnemonicToSecretKey(mnemonic);
  console.log(`📍 Deployer address: ${deployer.addr}`);

  // Check deployer balance
  const accountInfo = await algodClient.accountInformation(deployer.addr).do();
  const balance = accountInfo.amount / 1e6;
  console.log(`💰 Deployer balance: ${balance} ALGO\n`);

  if (balance < 1) {
    console.error('❌ Insufficient balance. Need at least 1 ALGO for deployment.');
    console.log('Get testnet ALGO from: https://bank.testnet.algorand.network/');
    process.exit(1);
  }

  // Read TEAL files
  const approvalPath = path.join(__dirname, '..', 'contracts', 'consent_approval.teal');
  const clearPath = path.join(__dirname, '..', 'contracts', 'consent_clear.teal');

  let approvalProgram, clearProgram;

  try {
    approvalProgram = fs.readFileSync(approvalPath, 'utf8');
    clearProgram = fs.readFileSync(clearPath, 'utf8');
  } catch (error) {
    console.error('❌ Could not read TEAL files. Run the Python compile scripts first:');
    console.log('   cd contracts && python approval.py && python clear.py');
    process.exit(1);
  }

  // Compile TEAL to bytecode
  console.log('📝 Compiling TEAL programs...');
  const approvalCompiled = await algodClient.compile(approvalProgram).do();
  const clearCompiled = await algodClient.compile(clearProgram).do();
  
  const approvalBytes = new Uint8Array(Buffer.from(approvalCompiled.result, 'base64'));
  const clearBytes = new Uint8Array(Buffer.from(clearCompiled.result, 'base64'));

  console.log(`   Approval program: ${approvalBytes.length} bytes`);
  console.log(`   Clear program: ${clearBytes.length} bytes\n`);

  // Get suggested params
  const suggestedParams = await algodClient.getTransactionParams().do();

  // Define schema
  const globalSchema = new algosdk.modelsv2.ApplicationStateSchema({
    numUint: 1,    // consent_counter
    numByteSlice: 1 // admin address
  });

  const localSchema = new algosdk.modelsv2.ApplicationStateSchema({
    numUint: 1,     // consent_count per user
    numByteSlice: 16 // up to 16 consent records per user
  });

  // Create application transaction
  console.log('📤 Creating application...');
  const txn = algosdk.makeApplicationCreateTxnFromObject({
    sender: deployer.addr,
    suggestedParams,
    approvalProgram: approvalBytes,
    clearProgram: clearBytes,
    numGlobalByteSlices: 1,
    numGlobalInts: 1,
    numLocalByteSlices: 16,
    numLocalInts: 1,
    onComplete: algosdk.OnApplicationComplete.NoOpOC,
  });

  // Sign and send
  const signedTxn = txn.signTxn(deployer.sk);
  const { txid } = await algodClient.sendRawTransaction(signedTxn).do();
  console.log(`   Transaction ID: ${txid}`);

  // Wait for confirmation
  console.log('⏳ Waiting for confirmation...');
  const confirmedTxn = await algosdk.waitForConfirmation(algodClient, txid, 4);
  
  const appId = confirmedTxn['application-index'];
  const appAddress = algosdk.getApplicationAddress(appId);

  console.log('\n✅ Deployment successful!');
  console.log('==================================');
  console.log(`📱 Application ID: ${appId}`);
  console.log(`📍 Application Address: ${appAddress}`);
  console.log(`🔗 View on explorer: https://testnet.algoexplorer.io/application/${appId}`);
  console.log(`📄 Transaction: https://testnet.algoexplorer.io/tx/${txid}`);

  // Save deployment info
  const deploymentInfo = {
    appId,
    appAddress,
    deployer: deployer.addr,
    txId: txid,
    network: 'testnet',
    timestamp: new Date().toISOString(),
  };

  const deploymentPath = path.join(__dirname, '..', 'deployment.json');
  fs.writeFileSync(deploymentPath, JSON.stringify(deploymentInfo, null, 2));
  console.log(`\n💾 Deployment info saved to: deployment.json`);

  return deploymentInfo;
}

// Run deployment
deployContract()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error('❌ Deployment failed:', error);
    process.exit(1);
  });
